# CHAPTER 03. 처음 시작하는 단위 테스트


## 3.1 환경 설정

제스트란? 목 객체(mock object)와 코드 커버리지(code coverage) 수집 기능까지 갖춘 메타의 오픈 소스

### 3.1.1 개발 환경 설정
### 3.1.2 예제 코드 저장소 구조
## 3.2 테스트 구성 요소
### 3.2.1 가장 단순한 테스트
### 3.2.2 테스트 구성 요소

```typescript
test(테스트명, 테스트 함수);
```

- 두 번째 인수인 테스트 함수에는 단언문을 작성함. = 검증문

```typescript
test("1 + 2는 3", ()=> {
    expect(검증값).toBe(기댓값);
})
```
- `expect()` 함수와 덧붙이는 매처(matcher)로 구성됨. 제스트는 공식적으로 여러 종류의 매처를 제공
- 단언문: expect(검증값).toBe(기댓값)
- 매처: toBe(기댓값)

### 3.2.3 테스트 그룹 작성

연관성 있는 테스트들을 그룹화하고 싶을 때는 `describe()` 사용

## 3.3 테스트 실행 방법
### 3.3.1 명령줄 인터페이스로 실행
### 3.3.2 제스트 러너로 실행
### 3.3.3 실행 결과 확인

## 3.4 조건 분기

사양이 복잡할 수록 조건 분기에서 버그가 많이 발생.

### 3.4.1 상한이 없는 덧셈 함수 테스트
### 3.4.2 하한이 있는 뺄셈 함수 테스트

## 3.5 에지 케이스와 예외 처리

예외 처리를 했는데 예상치 못한 입력값을 받을 경우 디버거debugger로 문제를 빨리 발견할 수 있음.

### 3.5.1 타입스크립트로 입력한 제약 설정

타입 애너테이션을 붙여서 다른 타입 값이 할당되면 실행전에 오류 발생시키기

### 3.5.2 예외 발생시키기

입력값이 기댓값과 다르면 예외를 발생시켜 값을 반환하기 전에 프로그램을 중지시킴.

### 3.5.3 예외 발생 검증 테스트

`expect(예외 발생 함수).toThrow()`

### 3.5.4 오류 메시지를 활용한 세부 사항 검증

의도치 않은 버그가 생겨서 예외가 발생할 때도 있음. 예외가 의도한 대로 발생했는지의 관점으로 접근하여 테스트코드를 작성하자.

### 3.5.5 instanceof 연산자를 활용한 세부 사항 검증

Error 클래스를 구체적인 상황에 맞춰 작성해 설계의 폭을 넓힐 수 있음.

```typescript
export class HttpError extends Error {}
export class RangeError extends Error {}

if (err instanceof HttpError) {
    //
}
if (err instanceof RangeError) {
    //
}
```

```typescript
function checkRange(value: number) {
    if(value < 0 || value > 100) {
        throw new RangError("0 ~ 100 사이의 값을 입력해주세요");
    }
}

export function add(a : number, b: number) {
    checkRang(a);
    checkRang(b);
    const sum = a + b;
    // ...
}
```

## 3.6 용도별 매처

### 3.6.1 진릿값 검증

`toBeTruthy`, `toBeFalsy`, `toBeNull`, `toBeUndefined`

### 3.6.2 수치 검증

`toBe`,`toBeCloseTo`

### 3.6.3 문자열 검증

`toContain`,`toMatch`, `toHaveLength`,`toEqual`, `stringContaining`,`stringMatching`

### 3.6.4 배열 검증

`toContain`,`toHaveLength`,`toContainEqual`,`arrayContaining`

### 3.6.5 객체 검증

`toMatchObject`,`toHaveProperty`, `objectContaining`

## 3.7 비동기 처리 테스트
### 3.7.1 테스트할 함수

경과시간을 반환값으로 resolve하는 함수

### 3.7.2 Promise를 반환하는 작성법

1. Promise를 반환하면서 then에 전달할 함수에 단언문을 작성하는 법.wait() 메서드 사용하면 Promise 인스턴스 생성됨. 해당 인스턴스를 테스트 함수의 반환 값으로 return 하면 Promise가 처리 중인 작업을 완료할 때까지 테스트 판정을 유예함.

2. resolves 매처를 사용하는 단언문을 return 하는 법.

### 3.7.3 async/await를 활용한 작성법

3. 테스트 함수를 async 함수로 만들어 함수 내에서 Promise가 완료될 때까지 기다리는 방법. 
4. 검증값인 Promise가 완료되는 것을 기다린 뒤 단언문을 실행하는 방법.

- async/await 함수를 사용하면 비동기 처리가 포함된 단언문이 여럿일 때 한 개의 테스트 함수 내에서 정리할 수 있는 장점이 있음.

### 3.7.4 Reject 검증 테스트

반드시 reject되는 함수를 사용해 reject된 경우를 검증하는 테스트로 예시를 듬.

1. Promise를 return하는 방법. catch메서드에 전달할 함수에 단언문을 작성함.
2. rejects 매처를 사용하는 단언문을 활용하는 방법. 단언문을 return 하거나 async/await을 사용함.
3. try-catch문 사용하는 방법. `UnhandledRejetion`을 try 블록에서 발생시키고 발생한 오류를 catch블록에서 받아 단언문으로 검증.

### 3.7.5 테스트 결과가 기댓값과 일치하는지 확인하기

실행하고 싶은 단언문에 도달하지 못한 채로 성공하며 종료되었을 경우, `expect.assertions`를 호출해야 함. 해당 메서드는 실행되어야 하는 단언문의 횟수를 인자로 받아 기대한 횟수만큼 단언문이 호출됐는지 검증함.


- 동기 함수라면 반드시 단언문을 return 해야함. 만약 return 을 빼먹지 않으려면 비동기 처리가 포함된 부분을 테스트할 때 아래와 같은 원칙을 가지고 접근하자.

> - 비동기 처리가 포함된 부분을 테스트할 때는 테스트 함수를 `async`함수로 만든다.
> - `.resolves`나`.rejects`가 포함된 단언문은 await한다.
> - `try-catch`문의 예외 발생을 검증할 때는 `expect.assertions`을 사용한다.